examples/pytorch/infograph/utils.py

```diff
@@ -0,0 +1,166 @@
+        Ep = -torch.exp(-p_samples)
+    elif measure == 'DV':
+        Ep = p_samples
+    elif measure == 'H2':
+        Ep = 1. - torch.exp(-p_samples)
+    elif measure == 'W1':
+        Ep = p_samples
+    else:
+        raise_measure_error(measure)
+
+    if average:
+        return Ep.mean()
+    else:
+        return Ep
+
+
+def get_negative_expectation(q_samples, measure, average=True):
+    """Computes the negative part of a divergence / difference.
+    Args:
+        q_samples: Negative samples.
+        measure: Measure to compute for.
+        average: Average the result over samples.
+    Returns:
+        torch.Tensor
+    """
+    log_2 = math.log(2.)
+
+    if measure == 'GAN':
+        Eq = F.softplus(-q_samples) + q_samples
+    elif measure == 'JSD':
+        Eq = F.softplus(-q_samples) + q_samples - log_2
+    elif measure == 'X2':
+        Eq = -0.5 * ((torch.sqrt(q_samples ** 2) + 1.) ** 2)
+    elif measure == 'KL':
+        Eq = torch.exp(q_samples)
+    elif measure == 'RKL':
+        Eq = q_samples - 1.
+    elif measure == 'DV':
+        Eq = log_sum_exp(q_samples, 0) - math.log(q_samples.size(0))
+    elif measure == 'H2':
+        Eq = torch.exp(q_samples) - 1.
+    elif measure == 'W1':
+        Eq = q_samples
+    else:
+        raise_measure_error(measure)
+
+    if average:
+        return Eq.mean()
+    else:
+        return Eq
+
+
+def local_global_loss_(l_enc, g_enc, graph_id, measure):
+    num_graphs = g_enc.shape[0]
+    num_nodes = l_enc.shape[0]
+
+    device = g_enc.device
+
+    pos_mask = th.zeros((num_nodes, num_graphs)).to(device)
+    neg_mask = th.ones((num_nodes, num_graphs)).to(device)
+
+    for nodeidx, graphidx in enumerate(graph_id):
+
+        pos_mask[nodeidx][graphidx] = 1.
+        neg_mask[nodeidx][graphidx] = 0.
+
+    res = th.mm(l_enc, g_enc.t())
+
+    E_pos = get_positive_expectation(res * pos_mask, measure, average=False).sum()
+    E_pos = E_pos / num_nodes
+    E_neg = get_negative_expectation(res * neg_mask, measure, average=False).sum()
+    E_neg = E_neg / (num_nodes * (num_graphs - 1))
+
+    return E_neg - E_pos
+
+
+def global_global_loss_(sup_enc, unsup_enc, measure):
+    '''
+    Args:
+        g: Global features
+        g1: Global features.
+        measure: Type of f-divergence. For use with mode `fd`
+        mode: Loss mode. Fenchel-dual `fd`, NCE `nce`, or Donsker-Vadadhan `dv`.
+    Returns:
+        torch.Tensor: Loss.
+    '''
+    num_graphs = sup_enc.shape[0]
+    device = sup_enc.device
+
+    pos_mask = th.eye(num_graphs).to(device)
+    neg_mask = 1 - pos_mask
+
+    res = th.mm(sup_enc, unsup_enc.t())
+
+    E_pos = get_positive_expectation(res * pos_mask, measure, average=False)
+    E_pos = (E_pos * pos_mask).sum() / pos_mask.sum()
+    E_neg = get_negative_expectation(res * neg_mask, measure, average=False)
+    E_neg = (E_neg * neg_mask).sum() / neg_mask.sum()
+
+    return E_neg - E_pos
+
+
+def adj_loss_(l_enc, g_enc, edge_index, batch):
```

Comment:
`batch` not used

Reply:
These functions are borrowed from the authors' repo. I am not sure whether I should make some modifications.